### ISSUE TITLE
Documentation for saving and reloading evaluations with Trials

### ISSUE THREAD

**Issue by bjkomer** [hyperopt/hyperopt#267]
Currently the wiki is not very clear that it is possible to save a set of evaluations and then continue where they were left off using the `Trials` object. It would be nice if a small example was added to the wiki that shows how to do this and mentions that the `max_evals` parameter refers to the total number of items in the trials database, rather than the number of evals to do for that specific `fmin` call.

There doesn't seem to be a nice way to do a pull request on a wiki, but I'm envisioning something like this:

```
import pickle
from hyperopt import fmin, tpe, hp, Trials, STATUS_OK

def objective(x):
    return {'loss': x ** 2, 'status': STATUS_OK }

# Initialize an empty trials database
trials = Trials()

# Perform 100 evaluations on the search space
best = fmin(objective,
    space=hp.uniform('x', -10, 10),
    algo=tpe.suggest,
    trials=trials,
    max_evals=100)

# The trials database now contains 100 entries, it can be saved/reloaded with pickle or another method
pickle.dump(trials, open("myfile.p", "wb"))
trials = pickle.load(open("myfile.p", "rb"))

# Perform an additional 100 evaluations
# Note that max_evals is set to 200 because 100 entries already exist in the database
best = fmin(objective,
    space=hp.uniform('x', -10, 10),
    algo=tpe.suggest,
    trials=trials,
    max_evals=200)

print(best)
```

**Comment by baziotis** [comment]
@bjkomer Very nice! I was looking for something like this. I have one question.

When we load a saved trials object and pass it to the `fmin`, does hyperopt take into account the previous runs? This is important in the case `algo=tpe.suggest`. Have you verified that?

**Comment by bjkomer** [comment]
@baziotis Good question! Yes, it does take previous runs into account. 

I just confirmed this on the toy example above by setting the `rstate` parameter in both the `fmin` calls, and then running it with and without the first `fmin` call, the final result is the exact same. In other words, whether you do two sets of 100 evaluations or one set of 200 evalutations, the result is the same.

I also did some digging in the code to be sure. [Line 156](https://github.com/hyperopt/hyperopt/blob/master/hyperopt/fmin.py#L156) of `fmin` is where the algorithm's `suggest` function is called and is given the `trials` object. This is the only way it takes into account previous runs, as `tpe.suggest` is just a function and has no internal state of its own. This function will return its suggested next place to look, which will then get appended to `trials` after being evaluated. This will keep looping until `max_evals` is reached.

I hope that helps!

**Comment by baziotis** [comment]
@bjkomer Very nice! :)
 > `tpe.suggest` is just a function and has no internal state of its own

That is all i wanted to hear. 

Also since this is the case here is what i do to run trials indefinitely and stop as soon as i want, without having to wait for all the `max_evals` to finish.

```python
def run_trials():

    trials_step = 1  # how many additional trials to do after loading saved trials. 1 = save after iteration
    max_trials = 5  # initial max_trials. put something small to not have to wait

    
    try:  # try to load an already saved trials object, and increase the max
        trials = pickle.load(open("my_model.hyperopt", "rb"))x
        print("Found saved Trials! Loading...")
        max_trials = len(trials.trials) + trials_step
        print("Rerunning from {} trials to {} (+{}) trials".format(len(trials.trials), max_trials, trials_step))
    except:  # create a new trials object and start searching
        trials = Trials()

    best = fmin(fn=mymodel, space=model_space, algo=tpe.suggest, max_evals=max_trials, trials=trials)

    print("Best:", best)
    
    # save the trials object
    with open(_model + ".hyperopt", "wb") as f:
        pickle.dump(trials, f)

# loop indefinitely and stop whenever you like
while True:
    run_trials()
```

Just let it keep on searching without having to worry that your progress will be lost or constraining yourself by having to wait for `max_trials`.

**Comment by guillaume-chevalier** [comment]
Thanks to all of you, your pieces of code has revealed useful to me in coding that small project: 
https://github.com/guillaume-chevalier/Hyperopt-Keras-CNN-CIFAR-100

However, I would now like to both add more parameters to the hyperparameters space AND resume the optimization from there. having already done a good search of the space. Would some of you have ideas on how to do that?

Here are some ideas that I have got, I would like to have feedback on that: 
- Simply try to resume the training and see if it works. Yet, I doubt that the TPE algorithm will be good at dealing with new hyperparameters it never saw upon resuming the training, but maybe it would work. 
- Edit the inner state of the `trials` variable before resuming training, to add my new hyperparameter: as if it was always to "false" before (the parameter I want to add is a hp.choice() of boolean).

**Comment by bjkomer** [comment]
@guillaume-chevalier I've never tried something like that myself, but I believe your second suggestion would work best, especially since you know a suitable value to set the old parameters to. Best of luck with your project!

**Comment by ericinlinux** [comment]
Super useful material! Thanks a lot for this @bjkomer! Going to use in my work as well.

**Comment by caugusta** [comment]
@cbaziotis Many thanks! I'm getting an error when I run your code. I think this line:

`with open(_model + ".hyperopt", "wb") as f:`

should instead be:

`with open("my_model.hyperopt", "wb") as f:`

Am I right, or does that change things badly?

**Comment by Codeguyross** [comment]
@cbaziotis 

I tried your example above but I am running into an issue with the pickle save on the trials variable. The error refers to a thread issue with pickling. Is anyone else experiencing this issue?

**Comment by bukhari-habib** [comment]
https://github.com/Codeguyross
are you using Keras ?

**Comment by bukhari-habib** [comment]
@Codeguyross 
Are you using sklearn or keras?

**Comment by jz314** [comment]
@cbaziotis I had the same problem and I think it's because it can't pickle the model object that is part of trials. My workaround was to delete it from trials.trials, which is a list of dictionaries, and then save:
```
for trial in trials.trials:
    del trial['result']['model'] # inplace deletion
# proceed with pickling
pickle.dump(trials, open("trials.pkl","wb"))
```

**Comment by Iamshg** [comment]
I think wiki should add the content .

**Comment by ninnghazad** [comment]
Trials() should accept a filename and do this on it's own.

**Comment by tobegit3hub** [comment]
Thanks @bjkomer and @cbaziotis .

Now we are trying to load the trials with only objective value instead of the `Trials` pickle. Is there any to to construct the `Trials` by pure Python code?

**Comment by antklen** [comment]
> another thing I noticed: when calling this via the hyperas wrapper using `tpe.suggest` mentioned [here](https://github.com/maxpumperla/hyperas/pull/102), it is important to make initial `max_evals` > 20 because the first 20 evals by default are random (code [here](https://github.com/hyperopt/hyperopt/blob/1a69042b208aba2191a97d7c1e0587ef84ffca46/hyperopt/tpe.py#L789)), and it appears if you resume from saved trials while it is still in the random sampling phase it forgets how many it has sampled and starts the random sequence over.
> You can use `partial` (example [here](https://github.com/hyperopt/hyperopt/blob/f38b538cfed9271e0fed5c5336eb8a45ae91d015/hyperopt/tests/test_pchoice.py#L121-L127)) to change the default number of random evals.

@jz314 Thanks for the trick with partial and n_startup_jobs. But could you elaborate more on this, why do you think that algorithm forgets how many it has sampled? And why it remembers if you make initial run with max_evals > n_startup_jobs? It's not clear to me from the code. From [here](https://github.com/hyperopt/hyperopt/blob/1a69042b208aba2191a97d7c1e0587ef84ffca46/hyperopt/tpe.py#L841) and [here](https://github.com/hyperopt/hyperopt/blob/1a69042b208aba2191a97d7c1e0587ef84ffca46/hyperopt/tpe.py#L869) it looks like it scans through the whole trials object and checks if it length < n_startup_jobs

**Comment by tangoed2whiskey** [comment]
@bjkomer 
> I just confirmed this on the toy example above by setting the `rstate` parameter in both the `fmin` calls, and then running it with and without the first `fmin` call, the final result is the exact same. In other words, whether you do two sets of 100 evaluations or one set of 200 evalutations, the result is the same.

I'm having trouble reproducing this; in the below commenting and uncommenting the code block changes the answer I'm getting at the end.  This warm-restart feature would be great, would it be possible to explain where I'm going wrong?

```
import pickle
from hyperopt import fmin, tpe, hp, Trials, STATUS_OK
import numpy as np

def objective(x):
    return {'loss': x ** 2, 'status': STATUS_OK }

# Initialize an empty trials database
trials = Trials()

# # Perform 50 evaluations on the search space
# best = fmin(objective,
#     space=hp.uniform('x', -10, 10),
#     algo=tpe.suggest,
#     trials=trials,
#     max_evals=50,
#     rstate=np.random.RandomState(314))

# The trials database now contains 100 entries, it can be saved/reloaded with pickle or another method
pickle.dump(trials, open("myfile.p", "wb"))
trials = pickle.load(open("myfile.p", "rb"))

# Perform an additional 50 evaluations
# Note that max_evals is set to 100 because 50 entries already exist in the database
best = fmin(objective,
    space=hp.uniform('x', -10, 10),
    algo=tpe.suggest,
    trials=trials,
    max_evals=100,
    rstate=np.random.RandomState(314))

print(best)
```

For concreteness, with the block uncommented I'm getting `{'x': -0.013171344359500809}` and with the block commented I'm getting `{'x': -0.004310266591624853}`

**Comment by bjkomer** [comment]
@tangoed2whiskey 

Very interesting... When I tested it I created a random state object by doing `my_rstate=np.random.randomState(314)` and then passing in `rstate=my_rstate` to each function call. When doing it that way it produces the same result whether or not the first block is commented out.

I believe the difference here is that in your example you are creating a new RandomState object each time, and this new state will not remember which random numbers where generated in the previous set of trials. This makes the two cases different and produce different results.

Seems like the fully correct thing to do when saving/loading trials would be to save the rstate as well, to make sure the same one is used in the subsequent evaluations.

**Comment by joelrich** [comment]
@guillaume-chevalier have you found a solution to the following problem (either with hyperopt or another library)?

> However, I would now like to both add more parameters to the hyperparameters space AND resume the optimization from there.

**Comment by guillaume-chevalier** [comment]
@joelrich I didn't try in the end. I ended up doing as follows: run a first hyperoptimization round with a huge domain. Then I narrow down the domain a lot and fix some parameters, and independently relaunch the optimization. By the way, I have a new repo that builds on the same code of the one I linked previously, and in this one I do such two hyperoptimization rounds: https://github.com/guillaume-chevalier/Linear-Attention-Recurrent-Neural-Network

By the way, I'd like someone to double-check what @jz314 said about the `tpe` trials being random or something.

**Comment by Paperone80** [comment]
Hi, thanks for all the comments and especially @tobegit3hub to get me on the right track. 

Here a complete example of how to fill a Trials() object based on a pd.Dataframe with previous trails and a corresponding hyperopt space.
Dataframe with 9 params and random 0|1:
```
df = pd.DataFrame(data=np.concatenate([np.random.choice([0,1], 90).reshape(10,9), np.random.choice([10,11], 10).reshape(10,1)], axis=1), columns=['param_'+str(i) for i in range(9)]+['loss'], index=range(10))
```

Define hyperopt space and check the sample:
```
hyperopt_space = {col: hyperopt.hp.choice(col, [0, 1]) for col in df.columns if col != 'loss'}
hyperopt_selection = hyperopt.pyll.stochastic.sample(hyperopt_space)
print(hyperopt_selection)
```

Now fill Trials(): 
```
test_trials = Trials()
for tid, (index, row) in enumerate(df.iterrows()):
    hyperopt_trial = hyperopt.Trials().new_trial_docs(
        tids=[tid],
        specs=[None], 
        results=[{'loss': row['loss'], 'status': hyperopt.STATUS_OK}],  
        miscs=[{'tid': tid
                , 'cmd': ('domain_attachment', 'FMinIter_Domain')
                , 'idxs': {**{key: [tid] for key in hyperopt_selection.keys()}}
                , 'vals': {**{key: [int(row[key])] for key in hyperopt_selection.keys()}}
                , 'workdir': None}]
       )
    hyperopt_trial[0]['state'] = hyperopt.JOB_STATE_DONE
    
    test_trials.insert_trial_docs(hyperopt_trial)   # [0] to return as dict{} from the list
    test_trials.refresh()

print(test_trials.trials)
```
Continue fmin():
```
best = fmin(
    <your function>,
    space=hyperopt_space,
    algo=tpe.suggest,
    max_evals=25,
    trials=test_trials
)
```

Hope that helps

**Comment by MilesCranmer** [comment]
Hi,
Does anyone have an example of slicing and merging two trial objects? It would be nice to have an easy way of combining separate optimization runs into a single trials object (this also offers a simple way of decentralized + distributed optimization), while also preventing duplicate trials. Thanks for the code @Paperone80 - I'll see if I can write this myself if nobody else has done this.
Cheers,
Miles

**Comment by MilesCranmer** [comment]
For anybody looking for code to merge two runs, this seems to merge trials objects succesfully! They require the same search space of course. Thanks @Paperone80 for the code example.

```python
import numpy as np
from hyperopt import Trials
import pickle as pkl
import datetime

trials = pkl.load(open('./bkuphyperopt_conv_log_100epoch_res17.pkl', 'rb'))
trials_new = pkl.load(open('./bkuphyperopt_conv_log_100epoch_res17.pkl', 'rb'))

def merge_trials(trials1, trials2_slice):
    max_tid = max([trial['tid'] for trial in trials1.trials])

    for trial in trials2_slice:
        tid = trial['tid'] + max_tid + 1
        hyperopt_trial = Trials().new_trial_docs(
                tids=[None],
                specs=[None],
                results=[None],
                miscs=[None])
        hyperopt_trial[0] = trial
        hyperopt_trial[0]['tid'] = tid
        hyperopt_trial[0]['misc']['tid'] = tid
        trials1.insert_trial_docs(hyperopt_trial) 
        trials1.refresh()
    return trials1

print(len(trials.trials))
#Add the last two runs from trials_new to trials.
t = merge_trials(trials, trials_new.trials[-2:])
print(len(t.trials))
```

**Comment by sebastienwood** [comment]
Could @MilesCranmer solution be integrated to the available out of the box methods of hyperopt ? I bet quite a few people would be interested to have it as a standard method.

**Comment by maxpumperla** [comment]
@MilesCranmer @sebastienwood sure, I could see that we merge this as util function in hyperopt. just need to make sure we have some safeguards with good error message when trials don't "match". Want to send a PR with some tests?

**Comment by PhilipMay** [comment]
Here is a nice tool for Trials persistence and restart of hyperopt: https://github.com/PhilipMay/mltb#module-hyperopt
For me it works very good.

**Comment by nd26** [comment]
@bjkomer 
Is it possible to load the trials if they were saved as follows (not pickled and saved):

```
domain = base.Domain(objective, search_space)
main_plot_history(trials)
data = pd.DataFrame(trials.trials)
data = pd.concat([data.drop(["result"], axis=1), data["result"].apply(pd.Series)], axis=1)
data = pd.concat([data.drop(["misc"], axis=1), data["misc"].apply(pd.Series)], axis=1)
data.sort_values(by=['loss'], ascending=False, inplace=True)
data.to_excel("Trials.xlsx")
```

**Comment by jnthnroy** [comment]
And the best will be the best of all trials or the best since we resume the optimization??

**Comment by guillaume-chevalier** [comment]
# A solution to all those problems

I'm working on a new library called [Neuraxle](https://github.com/Neuraxio/Neuraxle) to solve all of those trial saving and serialization problems. 

Not only that, but also we're adding data checkpoints such that if we've already seen some data, it can use cache to skip some preprocessing when some of the hyperparams are the same for the first preprocessing steps of a machine learning pipeline. 

## AutoMLWrapper

More precisely, an [AutoMLWrapper](https://github.com/Neuraxio/Neuraxle/issues/45) will be able to wrap those items and manage their interactions: 
- wrapped_pipeline : a ML model and its preprocessing and postprocessing steps. 
- auto_ml_strategy : Random Search, TPE, etc. Just inject (pass) an object to the wrapper. 
- validation_technique : either a Train/Validation wrapper, or a CrossValidation wrapper of the object
- score_function : a function to score the predicted against the expected
- hyperparams_repository : a place to save trials. It is abstract and a default implementation is provided. 
- n_iters : the budget. 

## Clean OOP structure

We plan on having different implementations of our wrapper such that it can try the trials sequentially, or in parallel, etc. This would therefore be compatible with Hyperband and so forth. The trick is to have clean enough Oriented Object Programming (OOP) such that everything is possible and compatible and customizeable out of abstract classes or interfaces to pass to the wrapper and whose can be easily derived. Although we plan to code this, we'd like contributors! :smile: 

## Hyperparameter spaces

We support hyperparameter spaces similar to `hp` spaces of hyperopt. See the [hyperparameter space definition tutorial](https://www.neuraxle.neuraxio.com/stable/hyperparameter_tuning.html) and a [full code example](https://www.neuraxle.neuraxio.com/stable/examples/boston_housing_meta_optimization.html#sphx-glr-examples-boston-housing-meta-optimization-py). 

## TPE

We'd actually appreciate to find a way to wrap Hyperopt into our wrapper, or code it from scratch again such that it finally have a clean OOP structure with inverted dependencies allowing for a custom trial saver (in our case, the `hyperparams_repository` as listed above, which is an abstract class open for extension). We aim at being able to easily pass any hyperparamater tuning algorithm into our wrappers, and be compatible with as much libraries.

**Comment by LucaUrbinati44** [comment]
I have a question for @bjkomer and @cbaziotis because I implemented your codes.

When the number of iterations arrives to be around 50, loading and saving with pickle becomes too slow (I'm talking about 30 minutes for a loading). I couldn't stand this overhead. So I tried saving the "trials" object with other methods: json, gzip and marshal. Every method told me the object "trials" is not in the proper format to be saved.

Moreover I tried to append (with open("my_model.hyperopt", "ab")) the most recent "trials" object to the file instead of saving all the "trials". Here the problem was that "trials" file became a list of dicts.

Thus, the question is: how did you manage the large overhead of saving/loading "trials"? Did you facing this problem like me or not?

What I'm doing now is to not use a function like run_trials(). I'm keeping "trials" as a global variable. In this way I can continue the computation when max_trials is reached. However the problem of saving/loading the "trials" object is only postponed.

Thank you

**Comment by PhilipMay** [comment]
@LucaUrbinati44 please have a look here: https://github.com/PhilipMay/mltb/blob/master/mltb/hyperopt.py

For me it works... fast and smooth.

**Comment by LucaUrbinati44** [comment]
@PhilipMay I tried your method using joblib with compression=1, 3 and 9. 
With compression=9 saving requires some seconds, while loading requires around 20 minutes for compression=1, 3 and 9. So I don't see any improvement with respect to pickle. 
(The test have been made with a "trials" object of 30 trials).

Do you have any other suggestions? Can it be a leak of RAM and not a software problem?

**Comment by bjkomer** [comment]
@jnthnroy yes, the best will be of all trials

> And the best will be the best of all trials or the best since we resume the optimization??

@nd26 unfortunately there is no built-in way to do that currently. I'm sure if someone has the time to make a PR it would get in. You will need to write your own code to fill a new trials object with your data. Take a look at the code posted by @Paperone80 and @MilesCranmer for ideas on how to do it.

> Is it possible to load the trials if they were saved as follows (not pickled and saved):
> 
> ```
> domain = base.Domain(objective, search_space)
> main_plot_history(trials)
> data = pd.DataFrame(trials.trials)
> data = pd.concat([data.drop(["result"], axis=1), data["result"].apply(pd.Series)], axis=1)
> data = pd.concat([data.drop(["misc"], axis=1), data["misc"].apply(pd.Series)], axis=1)
> data.sort_values(by=['loss'], ascending=False, inplace=True)
> data.to_excel("Trials.xlsx")
> ```

**Comment by PhilipMay** [comment]
> @PhilipMay I tried your method using joblib with compression=1, 3 and 9.
> With compression=9 saving requires some seconds, while loading requires around 20 minutes for compression=1, 3 and 9. So I don't see any improvement with respect to pickle.
> (The test have been made with a "trials" object of 30 trials).
> 
> Do you have any other suggestions? Can it be a leak of RAM and not a software problem?

Since you can store attachments in the trials object - see here: https://github.com/hyperopt/hyperopt/wiki/FMin#13-the-trials-object

It is possible that you add large things to it. We dont know about.

Also it is possible that you have a harddisk problem or use network attached storage?

Or it might be that you use very old software versions - is your software up to date? Do you use python 3.6 or 3.7?

**Comment by LucaUrbinati44** [comment]
I solved my problem: I was using [Hyperas](https://github.com/maxpumperla/hyperas) and not Hyperopt!

My goal was to save Hyperas "trials" object in the same way you did in this issue with Hyperopt. In fact I read here ([Continue optimization #217](https://github.com/maxpumperla/hyperas/issues/217)) that it could be possible to save the "trials" object. 
When I changed all my code to Hyperopt (thanks to this [examples](https://www.programcreek.com/python/example/98788/hyperopt.Trials)), loading and saving with pickle required me few milliseconds as @bjkomer stated two posts ago.

@bjkomer I'm not storing anything in my "trials". I directly used the "trials" object after the computation of optim.minimize() Hyperas function as you normally do after fmin() function in Hyperopt. What I saw is that after 30 evaluations the size of "my_model.pkl" was 27MB in Hyperas, while now it's just 4KB in Hyperopt.

Thanks to all the answers @PhilipMay, @bjkomer. I hope my problem can be useful for someone else, too.

**Comment by jnthnroy** [comment]
> @bjkomer Very nice! :)
> 
> > `tpe.suggest` is just a function and has no internal state of its own
> 
> That is all i wanted to hear.
> 
> Also since this is the case here is what i do to run trials indefinitely and stop as soon as i want, without having to wait for all the `max_evals` to finish.
> 
> ```python
> def run_trials():
> 
>     trials_step = 1  # how many additional trials to do after loading saved trials. 1 = save after iteration
>     max_trials = 5  # initial max_trials. put something small to not have to wait
> 
>     
>     try:  # try to load an already saved trials object, and increase the max
>         trials = pickle.load(open("my_model.hyperopt", "rb"))x
>         print("Found saved Trials! Loading...")
>         max_trials = len(trials.trials) + trials_step
>         print("Rerunning from {} trials to {} (+{}) trials".format(len(trials.trials), max_trials, trials_step))
>     except:  # create a new trials object and start searching
>         trials = Trials()
> 
>     best = fmin(fn=mymodel, space=model_space, algo=tpe.suggest, max_evals=max_trials, trials=trials)
> 
>     print("Best:", best)
>     
>     # save the trials object
>     with open(_model + ".hyperopt", "wb") as f:
>         pickle.dump(trials, f)
> 
> # loop indefinitely and stop whenever you like
> while True:
>     run_trials()
> ```
> 
> Just let it keep on searching without having to worry that your progress will be lost or constraining yourself by having to wait for `max_trials`.

Nice job!!! And how I can efficiently extract the best parameter set after reloading the trials??
Sorry Im in python only since one years and the trials give me some headache...

**Comment by PhilipMay** [comment]
> Nice job!!! And how I can efficiently extract the best parameter set after reloading the trials??
> Sorry Im in python only since one years and the trials give me some headache...

@jnthnroy You can call `trials.best_trial()` to get a dict with the best trial.

**Comment by jnthnroy** [comment]
> > Nice job!!! And how I can efficiently extract the best parameter set after reloading the trials??
> > Sorry Im in python only since one years and the trials give me some headache...
> 
> @jnthnroy You can call `trials.best_trial()` to get a dict with the best trial.
@PhilipMay Sorry my question was ot clear, I want to take a look on, for exemple top 10 best trials
put it inside a pandas with real values not the index of my space like I get with this command:

for t in range(1,len(trials.trials)):
    df = df.append([trials.trials[t]['misc']['vals']])

**Comment by christophelebrun** [comment]
Hello !

Trying to get the "best trial" from reloaded Trial(), I tried, as suggested, the following code :

```
import pickle

hypopt_trials = pickle.load(open("hp_trials/trials.hyperopt", "rb"))

hypopt_trials.best_trial()
```
Anyway, I get this error :
```
TypeError: 'dict' object is not callable
```

Any idea of how to fix it ?

@jnthnroy
You can get the actual parameters from index in the space using the syntax:
``` 
from hyperopt import space_eval
best_params = space_eval(space, best_trial)
```

**Comment by MilesCranmer** [comment]
I'm having issues merging large number of trials with the code I gave above. Sometimes I see the following issue when running fmin on merged trials:

```
    assert t_idxs == [] or t_idxs == [misc['tid']]
AssertionError
```

I'm not sure what causes it yet. But other than this, trial merging seems to work well.

**Comment by MilesCranmer** [comment]
Nevermind, I solved it. It's just because of more references to the trial ID in the trial docs. Here's an updated function for merging trials:

```python
def merge_trials(trials1, trials2_slice):
    max_tid = max([trial['tid'] for trial in trials1.trials])

    for trial in trials2_slice:
        tid = trial['tid'] + max_tid + 1
        hyperopt_trial = Trials().new_trial_docs(
                tids=[None],
                specs=[None],
                results=[None],
                miscs=[None])
        hyperopt_trial[0] = trial
        hyperopt_trial[0]['tid'] = tid
        hyperopt_trial[0]['misc']['tid'] = tid
        for key in hyperopt_trial[0]['misc']['idxs'].keys():
            hyperopt_trial[0]['misc']['idxs'][key] = [tid]
        trials1.insert_trial_docs(hyperopt_trial) 
        trials1.refresh()
    return trials1

trials_merged = merge_trials(trials, trials_new.trials[-2:])

```

**Comment by PhilipMay** [comment]
@MilesCranmer just out of curiosity: why do you want to merge trials?

**Comment by MilesCranmer** [comment]
@PhilipMay - To merge different optimization runs from separate nodes and different times

**Comment by MilesCranmer** [comment]
Here's the result: https://github.com/MilesCranmer/hyperopt_trials_merger.

It lets you use hyperopt in a distributed way with nothing but a shared folder between your processes. No head node, no distributed frameworks, and complete resilience to crashes of any process.

**Comment by MilesCranmer** [comment]
@maxpumperla Let me know if you want me to merge that code somehow. I think it's a really nice headache-free way to do distributed computing with nothing but a shared folder between your processes.

**Comment by hanan-vian** [comment]
will hyperopt work on offline data which was not generated by hyperopt trials?
e.g. offline warm-start?

**Comment by boomer-ai** [comment]
Hello,

Is it possible to rewrite a trials ? 

Imagine this scenario: you train a model and your terminal crashes for some reason, you want to resume the hyperopt from where it crashed, so you finish the training of the last model and write the trials by yourself with the hyperparameters of the model.

Will it take the trials into account for the next training as if no crash happened ?

Thank you for your answer

**Comment by ElrondL** [comment]
> ```python
> with open(savefile, 'w') as fid:
>     json.dump(trials.trials, fid, indent=4, sort_keys=True, default=str)
> ```

Hi, would json.dump() be run in inside the run_trials() function (i.e. updates every time the function is called)? Or can it just be a stand alone line?

**Comment by ElrondL** [comment]
Hi, what would happen if the hyper-parameter search space is altered before the trial is restarted? Is this advisable/will it induce TPE to perform badly? I know that this has kind of been asked before. TYVM for the explanations ^ ^

**Comment by ElrondL** [comment]
> Hi, what would happen if the hyper-parameter search space is altered before the trial is restarted? Is this advisable/will it induce TPE to perform badly? I know that this has kind of been asked before but would really appreciate a short analysis. I’m altering the parameters (value and type) within the first 20 trials, which would be where TPE does random searches.TYVM for the explanations ^ ^

Also, if the range of the parameter search spaces are changed, but the search items are not, would this be okay with TPE?

**Comment by ElrondL** [comment]
> Hi, what would happen if the hyper-parameter search space is altered before the trial is restarted? Is this advisable/will it induce TPE to perform badly? I know that this has kind of been asked before but would really appreciate a short analysis. I’m altering the parameters (value and type) within the first 20 trials, which would be where TPE does random searches.TYVM for the explanations ^ ^

@bjkomer
